rfcs/2021-03-26-6517-end-to-end-acknowledgement.md

```diff
@@ -0,0 +1,344 @@
+source can acknowledge them only after they are successfully delivered.
+
+## Scope
+
+This RFC will cover:
+
+* The functional requirements to support end-to-end acknowledgement
+
+* The data model changes and potential impacts to support the above
+
+
+This RFC will not cover:
+
+* Any details of changes to individual sources, transforms, or sinks to
+  necessary to support this feature.
+
+## Motivation
+
+Users want assurance that their data pipeline successfully delivers all the
+events that are presented to it. To accomplish this, Vector needs to be able to
+maintain information on each event tying it back to the source that originated
+the event, so that source can correctly acknowledge those incoming events at
+the right time.
+
+## Discussion
+
+Vector contains a number of features that can cause complications with a naïve
+implementation. As such, we need to itemize a number of scenarios to ensure our
+proposed solution covers all the edge cases. These will start with the simplest
+scenario and then introduce complicating factors.
+
+In the discussion, the term "finalization" is used to describe the point
+at which an event completes its trip through the pipeline, which may be
+caused by its acknowledged delivery, a permanent failure, or a situation
+causing the event to be dropped in processing.
+
+### Single source through linear transforms to a single sink
+
+This is the simplest of cases. For each event the source produces, the
+event is only modified on its transport through the system. The set of
+transforms represented here are all those not listed explicitly
+below. When the destination sink finishes handling the event, a
+notification needs to be delivered to the source indicating the
+finalization status. Thus, the event needs to contain a reference to the
+source and a unique message identifier. Different sources will have
+varying requirements on the message identifier.
+
+### Multiple sources to a single sink
+
+When multiple sources are configured to feed into a single sink, that sink may
+end up producing batches containing events from multiple sources. This does not
+really require any additional data tracking, but it is necessary to model the
+statuses as flowing back to the source independently for each event,
+not at a batch level.
+
+### Single source to multiple sinks
+
+If an event is to be sent to multiple sinks, we want the finalization status to
+only be issued once it has been finalized by all of the sinks. This will
+require the tracking be shared across all the relevant sinks (as opposed to
+being cloned with the event), as well as having a counter or some other
+mechanism for determining when all of the deliveries are completed. The
+topology will need to be involved in setting up this sink counter in order to
+avoid a race condition between handing the event off to the last sink and an
+earlier sink concluding that all deliveries are completed.
```

Comment:
I could see making the end-to-end acknowledgment configurable at the sink level too so that users could choose which sinks they want to have participate in acknowledgements. This could further be complicated by having multiple sources flowing into multiple sinks. For example, if I have sources A and B flowing into sinks C and D (both), I might want only A -> C to have end-to-end acks, but not B -> C or A -> D.